File: test_task_cli.py
import os
import json
import pytest
from task_cli import add, update, delete_task, mark_in, show_list, initialize_id, file_name, init

# Fixture for setup and teardown
@pytest.fixture(scope="function", autouse=True)
def setup_and_teardown():
    """
    Setup: Create a fresh JSON file before each test.
    Teardown: Delete the JSON file after each test.
    """
    if os.path.exists(file_name):
        os.remove(file_name)
    yield
    if os.path.exists(file_name):
        os.remove(file_name)

def load_tasks():
    """Helper function to load tasks from the JSON file."""
    with open(file_name, "r") as file:
        return json.load(file)["tasks"]

# Test cases
def test_initialize_id():
    assert initialize_id() == 1

def test_add_task():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("First Task", "todo")
    tasks = load_tasks()
    assert len(tasks) == 1, "A single task should be added"
    assert tasks[0]["description"] == "First Task", "Task description should match"
    assert tasks[0]["mark"] == "todo", "Task mark should be 'todo'"

def test_update_task():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task to Update", "todo")
    update(2, "Updated Task")
    tasks = load_tasks()
    assert tasks[0]["description"] == "Updated Task", "Task description should be updated"

def test_delete_task():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task to Delete", "todo")
    delete_task(3)
    tasks = load_tasks()
    assert len(tasks) == 0, "Task should be deleted successfully"

def test_mark_in_progress():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task to Mark", "todo")
    mark_in(4, "in-progress")
    tasks = load_tasks()
    assert tasks[0]["mark"] == "in-progress", "Task mark should be updated to 'in-progress'"

def test_mark_done():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task to Complete", "todo")
    mark_in(5, "done")
    tasks = load_tasks()
    assert tasks[0]["mark"] == "done", "Task mark should be updated to 'done'"

def test_show_list_all(capfd):
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task 1", "todo")
    add("Task 2", "in-progress")
    show_list("all")
    captured = capfd.readouterr()
    assert "Task 1" in captured.out, "Task 1 should be displayed"
    assert "Task 2" in captured.out, "Task 2 should be displayed"

def test_show_list_filtered(capfd):
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task 1", "todo")
    add("Task 2", "in-progress")
    show_list("todo")
    captured = capfd.readouterr()
    assert "Task 1" in captured.out, "Filtered list should include Task 1"
    assert "Task 2" not in captured.out, "Filtered list should exclude Task 2"

def test_nonexistent_task_update():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    update(99, "Nonexistent Update")
    tasks = load_tasks()
    assert len(tasks) == 0, "No tasks should be modified for a nonexistent task"

def test_nonexistent_task_delete():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    delete_task(99)
    tasks = load_tasks()
    assert len(tasks) == 0, "No tasks should be deleted for a nonexistent task"

def test_nonexistent_task_mark():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    mark_in(99, "done")
    tasks = load_tasks()
    assert len(tasks) == 0, "No tasks should be marked for a nonexistent task"

def test_add_multiple_tasks():
    data = {"tasks": []}
    with open(file_name, 'w') as file:
            json.dump(data, file, indent=4)
    initialize_id()
    add("Task 1", "todo")
    add("Task 2", "todo")
    tasks = load_tasks()
    assert len(tasks) == 2, "Two tasks should be added"
    assert tasks[0]["description"] == "Task 1", "First task description should match"
    assert tasks[1]["description"] == "Task 2", "Second task description should match"

def test_update_marked_task():
    initialize_id()
    add("Task to Mark and Update", "todo")
    mark_in(12, "in-progress")
    update(12, "Updated and Marked Task")
    tasks = load_tasks()
    print(tasks)
    assert tasks[0]["description"] == "Updated and Marked Task", "Description should be updated"
    assert tasks[0]["mark"] == "in-progress", "Mark should remain 'in-progress'"
# No tasks found in JSON file. Using default ID = 1.
# New task added: {'ID': 12}
# Task ID 1 not found.
# Task ID 1 not found.
# [{'description': 'Task to Mark and Update', 'ID': 12, 'mark': 'todo'}]

# Main execution
if __name__ == "__main__":
    pytest.main(["-v", "--tb=short"])

File: cpp/test_task_cli.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"  // Include doctest header
// #include "task_cli.hpp" for makefile
#include "task_cli.cpp"  // Include your task CLI implementation
#include <fstream>
#include <nlohmann/json.hpp>
#include <string>

using json = nlohmann::json;
using namespace std;

// const string file_name = "tasks.json";
void print_json_file(const std::string& filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Could not open the file " << filename << std::endl;
        return;
    }

    try {
        nlohmann::json j;
        file >> j;
        std::cout << "Contents of " << filename << ":\n" << j.dump(4) << std::endl;  // Pretty print with indentation
    } catch (const std::exception& e) {
        std::cerr << "Error parsing JSON from " << filename << ": " << e.what() << std::endl;
    }
}
// Helper functions for setup and teardown
void setup_file() {
    ifstream file(file_name);
    if (!file.is_open()) {
        json data = {{"tasks", json::array()}};
        ofstream out_file(file_name);
        out_file << setw(4) << data;
        out_file.close();
    }
}

void teardown_file() {
    if (remove(file_name.c_str()) != 0) {
        perror("Error deleting file");
    }
}

json load_tasks() {
    ifstream file(file_name);
    json data;
    file >> data;
    return data["tasks"];
}

// Test cases
TEST_CASE("Initialize ID starts at 1") {
    teardown_file();
    initialize_id();
    CHECK(id == 1);
}

TEST_CASE("Add a task successfully") {
    setup_file();
    initialize_id();
    add_task("First Task", "todo");
    json tasks = load_tasks();
    CHECK(tasks.size() == 1);
    CHECK(tasks[0]["description"] == "First Task");
    CHECK(tasks[0]["mark"] == "todo");
}

TEST_CASE("Update a task description") {
    setup_file();
    initialize_id();
    add_task("Task to Update", "todo");
    update_task(1, "Updated Task");
    json tasks = load_tasks();
    CHECK(tasks[0]["description"] == "Updated Task");
}

TEST_CASE("Delete a task") {
    setup_file();
    initialize_id();
    add_task("Task to Delete", "todo");
    delete_task(3);
    json tasks = load_tasks();
    CHECK(tasks.size() == 2);
}

TEST_CASE("Mark a task as in-progress") {
    setup_file();
    initialize_id();
    add_task("Task to Mark", "todo");
    mark_task(1, "in-progress");
    json tasks = load_tasks();
    CHECK(tasks[0]["mark"] == "in-progress");
}

TEST_CASE("Mark a task as done") {
    setup_file();
    initialize_id();
    add_task("Task to Complete", "todo");
    mark_task(1, "done");
    json tasks = load_tasks();
    CHECK(tasks[0]["mark"] == "done");
}

TEST_CASE("Show all tasks") {
    setup_file();
    initialize_id();
    add_task("Task 1", "todo");
    add_task("Task 2", "in-progress");

    stringstream buffer;
    streambuf* old = cout.rdbuf(buffer.rdbuf());
    show_list("all");
    cout.rdbuf(old);

    string output = buffer.str();
    CHECK(output.find("Task 1") != string::npos);
    CHECK(output.find("Task 2") != string::npos);
}

TEST_CASE("Show filtered tasks") {
    setup_file();
    initialize_id();
    add_task("Task 1", "todo");
    add_task("Task 2", "in-progress");

    stringstream buffer;
    streambuf* old = cout.rdbuf(buffer.rdbuf());
    show_list("todo");
    cout.rdbuf(old);

    string output = buffer.str();
    CHECK(output.find("Task 1") != string::npos);
    CHECK(output.find("Task 2") == string::npos);
}

TEST_CASE("Handle update of a nonexistent task") {
    setup_file();
    initialize_id();
    update_task(99, "Nonexistent Task");
    json tasks = load_tasks();
    // print_json_file(file_name);
    CHECK(tasks.size() == 8);
}

TEST_CASE("Handle deletion of a nonexistent task") {
    setup_file();
    initialize_id();
    delete_task(99);
    json tasks = load_tasks();
    // print_json_file(file_name);
    CHECK(tasks.size() == 8);
}

TEST_CASE("Handle marking a nonexistent task") {
    setup_file();
    initialize_id();
    mark_task(99, "done");
    json tasks = load_tasks();
    // print_json_file(file_name);
    CHECK(tasks.size() == 8);
}

TEST_CASE("Add multiple tasks") {
    setup_file();
    initialize_id();
    add_task("Task 1", "todo");
    add_task("Task 2", "todo");
    json tasks = load_tasks();
    // print_json_file(file_name);
    CHECK(tasks.size() == 10);
    CHECK(tasks[8]["description"] == "Task 1");
    CHECK(tasks[9]["description"] == "Task 2");
}

TEST_CASE("Update a marked task") {
    setup_file();
    initialize_id();
    add_task("Task to Update and Mark", "todo");
    mark_task(11, "in-progress");
    update_task(11, "Updated Task");
    json tasks = load_tasks();
    // print_json_file(file_name);
    CHECK(tasks[10]["description"] == "Updated Task");
    CHECK(tasks[10]["mark"] == "in-progress");
}
